src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,184 @@
+	if found {
+		// Credentials found in cache
+		credentials = res.(models.CustomMetricCredentials)
+		isValid = mh.validateCredentials(username, credentials.Username, password, credentials.Password)
+	}
+
+	// Credentials not found in cache or
+	// stale cache entry with invalid credential found in cache
+	// search in the database and update the cache
+	if !found || !isValid {
+		usernameHash, passwordHash, err := mh.policyDB.GetCustomMetricsCreds(appID)
+		if err != nil {
+			mh.logger.Error("error-during-getting-binding-credentials-from-policyDB", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+				Code:    "Interal-Server-Error",
+				Message: "Error reading custom metrics request body"})
+			return
+		}
+		credentials = models.CustomMetricCredentials{
+			Username: usernameHash,
+			Password: passwordHash,
+		}
+		isValid = mh.validateCredentials(username, credentials.Username, password, credentials.Password)
+		// If Credentials in DB is not valid
+		if !isValid {
+			mh.logger.Error("error-validating-authorizaion-header", err, lager.Data{"authheader": auth})
+			http.Error(w, "Authorization failed", http.StatusUnauthorized)
+			handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+				Code:    "Authorization-Failure-Error",
+				Message: "Basic authorization credential does not match"})
+			return
+		}
+		// Valid credentials found in DB so update the cache
+		mh.credentialCache.Set(appID, credentials, mh.cacheTTL)
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		mh.logger.Error("error-reading-request-body", err, lager.Data{"body": r.Body})
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error reading custom metrics request body"})
+		return
+	}
+	var metricsConsumer *models.MetricsConsumer
+	err = json.Unmarshal(body, &metricsConsumer)
+	if err != nil {
+		mh.logger.Error("error-unmarshaling-metrics", err, lager.Data{"body": r.Body})
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error unmarshaling custom metrics request body"})
+		return
+	}
+	isValidRequest, err := mh.policyDB.ValidateCustomMetricTypes(appID, metricsConsumer)
+	if !isValidRequest && err != nil {
+		mh.logger.Error("failed-validating-metrictypes", err, lager.Data{"metrics": metricsConsumer})
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
+			Code:    "Bad-Request",
+			Message: "Error validating custom metrics type"})
+		return
+	}
+
+	metrics := mh.parseMetrics(metricsConsumer)
+
+	if len(metrics) <= 0 {
```

Comment:
how about moving this len() check prior to ValidateCustomMetricTypes()?   I assume Len(metrics) ==0 only happens when len(metricsConsumer.CustomMetrics) ==0

Reply:
I still think we need to do  len(metrics) <= 0 check first .  moving this part inside ValidateCustomMetricTypes(), and make it happen before checking the allowed metric types will save some kinds of unnecessary computation.

Reply:
I would agree, while you can just check whether len(metricsConsumer.CustomMetrics) == 0 or not.